Catty-188: Refactor Brick tests

neel-makhecha opened the pull request:
Added new tests to test 'mutable copy' functionality for various bricks. Removed unnecessary overriding 'mutableCopy' methods and added it where required in various subclasses of Brick. 

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catty/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [ ] Include the name of the Jira ticket in the PR’s title
- [ ] Verify that the Jira ticket is in the status *Ready for Development*
- [ ] Include a summary of the changes plus the relevant context
- [ ] Choose the proper base branch (*develop*)
- [ ] Confirm that the changes follow the project’s coding guidelines
- [ ] Verify that the changes generate no compiler or linter warnings
- [ ] Perform a self-review of the changes
- [ ] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Confirm that new and existing unit tests pass locally
- [ ] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *#catty* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

m-herold commented on src/Catty/DataModel/Bricks/Control/RepeatUntilBrick.m:
```diff
@@ -69,13 +69,6 @@ - (NSString*)description
 }
 
-#pragma mark - Copy
-- (id)mutableCopyWithContext:(CBMutableCopyContext*)context
```
The _RepeatUntilBrick_ contains only one Formula. When copying a Brick, there should also be a copy of the Formula, so that the Formula can be edited separately (in the original and the copied Brick). This happens all in the parent method, so no overriding necessary. Thanks!

m-herold commented on src/Catty/DataModel/Bricks/Control/WaitUntilBrick.m:
```diff
@@ -69,13 +69,6 @@ - (NSString*)description
 }
 
-#pragma mark - Copy
-- (id)mutableCopyWithContext:(CBMutableCopyContext*)context
```
See comment above.

m-herold commented on src/Catty/DataModel/Bricks/Look/SetBackgroundBrick.m:
```diff
@@ -78,21 +78,4 @@ - (NSInteger)getRequiredResources
 }
 
-
-- (id)mutableCopyWithContext:(CBMutableCopyContext*)context
```
Different to Bricks containing Formulas, Looks should not be copied when a Brick gets copied. Looks are Object-wide assets. So a copied _SetBackgroundBrick_ within an Object should still point to the same Look within that Object. Exception: when copying the whole Object, this reference needs to be updated. Therefore I think we still need that method -- probably some of the _SpriteObjectMutableCopyTests_ will be failing. Please add a separate test to your _SetBackgroundBrickTests_ checking this behavior by making use of the _CBMutableCopyContext_.

neel-makhecha replied:
Got the point! Upon adding the previously removed method, I think I should also add another override of that method which have reportError bool argument. Otherwise, those calls will be redirected to parent method.

m-herold replied:
Thanks! And good point. The interface of the Brick is misleading since the other method could be called as well, right. However, this may cause a few more changes in this pr which makes review even more difficult, so in this ticket please only stick to the _mutableCopyWithContext_ method like the one you deleted here. I will create a separate ticket for the concern you mentioned which will deal with this issue.

m-herold commented on src/Catty/DataModel/Bricks/Look/SetLookBrick.m:
```diff
@@ -83,21 +83,4 @@ - (NSInteger)getRequiredResources
 }
 
-
-- (id)mutableCopyWithContext:(CBMutableCopyContext*)context
```
Same applies here, see comment above.

m-herold commented on src/Catty/DataModel/Bricks/Motion/GlideToBrick.m:
```diff
@@ -103,13 +103,6 @@ - (BOOL)isEqualToBrick:(Brick*)brick
 }
 
-#pragma mark - Copy
-- (id)mutableCopyWithContext:(CBMutableCopyContext*)context
```
Looks good to me!

m-herold commented on src/Catty/DataModel/Bricks/Motion/PointToBrick.m:
```diff
@@ -84,6 +84,14 @@ - (id)mutableCopyWithContext:(CBMutableCopyContext*)context
     return copy;
 }
 
+- (id)mutableCopyWithContext:(CBMutableCopyContext*)context AndErrorReporting:(BOOL)reportError
```
Can you give me some hints why you added this method?

neel-makhecha replied:
The override of method with two arguments (reportError Boolean) wasn't present. So calls made by also passing the boolean for reportError always went to the parent method. Hence, added this argument.

m-herold commented on src/Catty/DataModel/Bricks/Sound/PlaySoundAndWaitBrick.m:
```diff
@@ -31,23 +31,6 @@ - (kBrickCategoryType)category
 }
 
-#pragma mark - Copy
-- (id)mutableCopyWithContext:(CBMutableCopyContext*)context
```
Similar to the "Look Bricks", the same applies to the "Sound Bricks". Please refer to my previous comment about the _SetBackgroundBrick_.

m-herold commented on src/CattyTests/Bricks/BrickTests.swift:
```diff
@@ -67,4 +67,34 @@ final class BrickTests: XCTestCase {
                        "Invalid formula")
     }
+
+    func testMutableCopy() {
```
Looks good! However, in the meantime this has been resolved in another (training) ticket. Sorry for the confusion, this should have been mentioned in your ticket. Please remove your two tests from this class. Additional, you could improve the existing tests in _BrickMutableCopyTests_ by adding the assertions for references as suggested below.

m-herold commented on src/CattyTests/Bricks/GlideToBrickTests.swift:
```diff
@@ -37,4 +37,40 @@ final class GlideToBrickTests: XCTestCase {
+        XCTAssertTrue(brick.yDestination.isEqual(to: copiedBrick.yDestination), "yDestination properties are equal")
+    }
+
+    func testMutableCopyForBool() {
```
The _isAnimated_ and _isAnimatedInsertBrick_ properties are some inherited burdens which actually should not be publicly visible. They need some major refactoring, but that will be handled in the future. Since the copying of these properties are handled in other tests (e.g. _BrickMutableCopyTests_), there is no need to retest this functionality (I would also suggest to not separately document it in every test). Please remove these tests in your new Brick test classes.

m-herold commented on src/CattyTests/Bricks/GlideToBrickTests.swift:
```diff
@@ -37,4 +37,40 @@ final class GlideToBrickTests: XCTestCase {
+        brick.xDestination = Formula(double: 1)
+        brick.yDestination = Formula(double: 1)
+
+        let copiedBrick: GlideToBrick = brick.mutableCopy(with: CBMutableCopyContext(), andErrorReporting: true) as! GlideToBrick
```
Since we want to make absolutely sure that the copied Brick is not only equal to the original Brick, but also references a separate instance, I would suggest adding one more assert for each copiedBrick, e.g. something like _XCTAssertFalse(brick === copiedBrick)_.

m-herold commented on src/CattyTests/Bricks/GlideToBrickTests.swift:
```diff
@@ -37,4 +37,40 @@ final class GlideToBrickTests: XCTestCase {
+
+        XCTAssertTrue(brick.isEqual(to: copiedBrick), "GlideToBricks are not equal")
+        XCTAssertTrue(brick.durationInSeconds.isEqual(to: copiedBrick.durationInSeconds), "durationInSeconds properties are equal")
+        XCTAssertTrue(brick.xDestination.isEqual(to: copiedBrick.xDestination), "xDestination properties are equal")
```
The same is even more important for the formulas contained in the Bricks. please also add something like that: `XCTAssertFalse(brick.xDestination === copiedBrick.xDestination)`

m-herold commented on src/CattyTests/Bricks/IfLogicBeginBrickTests.swift:
```diff
@@ -0,0 +1,56 @@
+
+@testable import Pocket_Code
+
+final class IfLogicBeginBrickTests: XCTestCase {
```
See my comments above.

m-herold commented on src/CattyTests/Bricks/PlaySoundAndWaitBrickTests.swift:
```diff
@@ -0,0 +1,58 @@
+
+@testable import Pocket_Code
+
+final class PlaySoundAndWaitBrickTests: XCTestCase {
```
See my comments above.

m-herold commented on src/CattyTests/Bricks/PointToBrickTests.swift:
```diff
@@ -82,4 +82,26 @@ final class PointToBrickTests: AbstractBrickTest {
         XCTAssertEqual(45.0, firstSpriteNode.catrobatRotation, accuracy: 0.1, "PointToBrick not correct")
     }
 
+    func testMutableCopy() {
```
See my comments above.

m-herold commented on src/CattyTests/Bricks/RepeatBrickTests.swift:
```diff
@@ -73,4 +73,41 @@ final class RepeatBrickTests: XCTestCase {
         XCTAssertNil(repeatBrick.maxRepetitions)
     }
+
+    func testMutableCopy() {
```
See my comments above.

m-herold commented on src/CattyTests/Bricks/RepeatUntilBrickTests.swift:
```diff
@@ -49,4 +49,37 @@ final class RepeatUntilBrickTests: XCTestCase {
         XCTAssertFalse(brick.checkCondition(formulaInterpreter: formulaInterpreter))
     }
+
+    func testMutableCopy() {
```
See my comments above.

m-herold commented on src/CattyTests/Bricks/SetBackgroundBrickTests.swift:
```diff
@@ -26,13 +26,26 @@ import XCTest
-        let object = SpriteObject()
-        let project = Project.defaultProject(withName: "a", projectID: "1")
-        let spriteNode = CBSpriteNode(spriteObject: object)
+    var object: SpriteObject!
```
Nice, thanks!

m-herold commented on src/CattyTests/Bricks/SetBackgroundBrickTests.swift:
```diff
@@ -69,4 +80,59 @@ final class SetBackgroundBrickTests: AbstractBrickTest {
         Project.removeProjectFromDisk(withProjectName: project.header.programName, projectID: project.header.programID)
     }
+
+    func testMutableCopy() {
```
See my comments above.

m-herold commented on src/CattyTests/Bricks/SetBackgroundBrickTests.swift:
```diff
@@ -69,4 +80,59 @@ final class SetBackgroundBrickTests: AbstractBrickTest {
+
+        let bundle = Bundle(for: type(of: self))
+        let filePath = bundle.path(forResource: "test.png", ofType: nil)
+        let imageData = UIImage(contentsOfFile: filePath!)!.pngData()
```
The actual image data is not necessary in this test and can be removed.

m-herold commented on src/CattyTests/Bricks/SetLookBrickTests.swift:
```diff
@@ -0,0 +1,94 @@
+
+    }
+
+    func testMutableCopy() {
```
See my comments above.

m-herold commented on src/CattyTests/Bricks/SetLookBrickTests.swift:
```diff
@@ -0,0 +1,94 @@
+    func testMutableCopy() {
+        let bundle = Bundle(for: type(of: self))
+        let filePath = bundle.path(forResource: "test.png", ofType: nil)
+        let imageData = UIImage(contentsOfFile: filePath!)!.pngData()
```
Again here, I think the image data is not needed.

m-herold replied:
Is the _bundle_, _filePath_ and _imageData_ really necessary?

m-herold commented on src/CattyTests/Bricks/WaitUntilBrickTests.swift:
```diff
@@ -80,6 +80,38 @@ final class WaitUntilBrickTests: XMLAbstractTest {
         XCTAssertFalse(conditionResult, "Condition should have returned false.")
     }
 
+    func testMutableCopy() {
```
See my comments above.

m-herold commented on src/CattyTests/Bricks/GlideToBrickTests.swift:
```diff
@@ -37,4 +37,40 @@ final class GlideToBrickTests: XCTestCase {
+
+        let copiedBrick: GlideToBrick = brick.mutableCopy(with: CBMutableCopyContext(), andErrorReporting: true) as! GlideToBrick
+
+        XCTAssertTrue(brick.isEqual(to: copiedBrick), "GlideToBricks are not equal")
```
You tests are very readable. Please remove all messages in the assert statements since they do not deliver any additional information (your test code explains already everything).

neel-makhecha commented:
> Added new tests to test 'mutable copy' functionality for various bricks. Removed unnecessary overriding 'mutableCopy' methods and added it where required in various subclasses of Brick.
> 
> ### Your checklist for this pull request
> Please review the [contributing guidelines](https://github.com/Catrobat/Catty/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.
> 
> * [ ]  Include the name of the Jira ticket in the PR’s title
> * [ ]  Verify that the Jira ticket is in the status _Ready for Development_
> * [ ]  Include a summary of the changes plus the relevant context
> * [ ]  Choose the proper base branch (_develop_)
> * [ ]  Confirm that the changes follow the project’s coding guidelines
> * [ ]  Verify that the changes generate no compiler or linter warnings
> * [ ]  Perform a self-review of the changes
> * [ ]  Verify to commit no other files than the intentionally changed ones
> * [ ]  Include reasonable and readable tests verifying the added or changed behavior
> * [ ]  Confirm that new and existing unit tests pass locally
> * [ ]  Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
> * [ ]  Stick to the project’s git workflow (rebase and squash your commits)
> * [ ]  Verify that your changes do not have any conflicts with the base branch
> * [ ]  After the PR, verify that all CI checks have passed
> * [ ]  Post a message in the _#catty_ [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

neel-makhecha commented:
> Thanks, your pull request looks clear and concise. I have added a few comments directly to the affected lines of code.

I have made changes as suggested by you. Please review them.

m-herold commented on src/CattyTests/Bricks/RepeatBrickTests.swift:
```diff
@@ -73,4 +73,27 @@ final class RepeatBrickTests: XCTestCase {
+        XCTAssertFalse(brick === copiedBrick)
+        XCTAssertEqual(brick.maxRepetitions, copiedBrick.maxRepetitions)
+        XCTAssertEqual(copiedBrick.repetitions, 1)
+        XCTAssertTrue(brick.maxRepetitions === copiedBrick.maxRepetitions)
```
The _maxRepetitions_ and _repetitions_ instances should not be the same references.

neel-makhecha replied:
Thanks for confirming that! But the test is failing. Earlier, I had removed the mutableCopy override from RepeatBrick, which I added back and tested too. But still the test is failing, i.e. references for maxRepetitions are getting same regardless there's an override or not. 
In fact, overriding method which I had removed had a following statement before returning the brick...
brick.maxRepetitions = self.maxRepetitions;
I think this statement will actually make reference of maxRepetitions same.

m-herold replied:
Thanks! I think bringing back that method is a good idea. Please try to initialize them with zero, this will actually be more accurate since the newly copied Brick should have its own, independent "repetition logic".

neel-makhecha replied:
I think what I understand is I should be comparing timesToRepeat instead of maxRepetitions and repetitions.

m-herold commented on src/CattyTests/Bricks/RepeatUntilBrickTests.swift:
```diff
@@ -49,4 +49,26 @@ final class RepeatUntilBrickTests: XCTestCase {
+        brick.repeatCondition = Formula.init(double: 0)
+
+        XCTAssertFalse(brick.repeatCondition.isEqual(to: copiedBrick.repeatCondition))
+        XCTAssertFalse(brick.repeatCondition === copiedBrick)
```
Please revise this statement.

neel-makhecha replied:
I am sorry about the mistake. But I think I should remove this test as the values are already being tested. If that isn't equal, then references won't be equal anyway.

m-herold replied:
I agree. Please do not remove the whole test, but just this assert statement

neel-makhecha commented:
Hi Michael,
Just made a commit with latest changes. Please review it.

m-herold commented:
Thanks a lot!

m-herold commented:
Everything looks good. Please stick to the project guidelines and squash your commit messages to one commit with a meaningful commit message (i.e. the title of your PR which is perfect). Let me know once you have done this, so that I can directly merge your PR.

neel-makhecha commented:
Hi Michael,
I have squashed all commits to one as per guidelines. Please review it.

m-herold commented:
Jenkins, retest this please.

neel-makhecha commented:
Hi Michael,
Is there any problem with the last commit?
I don't understand.

m-herold commented:
Not at all, everything looks fine! The Jenkins master is currently having some organization-wide problems. I will retrigger the build and merge your pr once the master is up and running again.
